docs/guides/01_creating_an_aggregate.md

```diff
@@ -0,0 +1,190 @@
+
+    private Task() { }
+
+    public static Task Create(string name, string email)
+    {
+        return new Task
+        {
+            Id = SId<User>.New(),
+            Name = name,
+            Email = email,
+        };
+    }
+}
+```
+
+Notice that the `User` class is an aggregate root - the user represents a standalone business domain object with its own meaning and behavior.
+
+## Adding logic
+
+So far, only the structure of objects has been defined but business domain objects usually has some set of behaviors which they can perform. In this section we will see how a logic can be added to the aggregates.
+
+Let's start with a simple example of adding a few methods to the `Task` class:
+
+```csharp
+public class Task : IIdentifiable<SId<Task>>
+{
+    . . .
+
+    public void Edit(string name)
+    {
+        Name = name;
+    }
+
+    public void AssignUser(SId<User> userId)
+    {
+        AssignedUser = userId;
+    }
+
+    public void UnassignUser()
+    {
+        AssignUser = null;
+    }
+
+    public void ChangeStatus(TaskStatus status)
+    {
+        Status = status;
+    }
+}
+```
+
+Remember that the only way to access a task is to go through the relevant project which is the aggregate root. Because of this we will add some methods to the `Project` class which will allow us to interact with the tasks:
+
+```csharp
+public class Project : IIdentifiable<SId<Project>>
+{
+    . . .
+
+    public void EditTask(SId<Task> taskId, string name)
+    {
+        tasks.Single(t => t.Id == taskId).Edit(name);
+    }
+
+    public void AssignUserToTask(SId<Task> taskId, SId<User> userId)
+    {
+        tasks.Single(t => t.Id == taskId).AssignUser(userId);
+    }
+
+    public void UnassignUserFromTask(SId<Task> taskId)
+    {
+        tasks.Single(t => t.Id == taskId).UnassignUser(userId);
+    }
+
+    public void ChangeTaskStatus(SId<Task> taskId, TaskStatus status)
+    {
+        tasks.Single(t => t.Id == taskId).ChangeTaskStatus(status);
+    }
+}
+```
+
+Notice that these added methods can and will throw an exception if a project does not contain any task with provided Id. This is an excepted behavior - checks for respecting domain logic should be performed in respective command (or operations) validators.
+
+## Rasing domain events
+
+The domain events make an important part of DDD. It is through the use of domain events that an aggregate might communicate with the rest of the system. We will see how to raise an event from the aggregate.
+
+Let's imagine that we want to perform some action after a user has been assigned to a task. We will modify `AssignUserToTask` method from the `Project` class:
+
+```csharp
+public class Project : IIdentifiable<SId<Project>>
+{
+    . . .
+
+    public void AssignUserToTask(SId<Task> taskId, SId<User> userId)
+    {
+        tasks.Single(t => t.Id == taskId).AssignUser(userId);
+        DomainEvents.Raise(new UserAssignedToTask(SId<Task> taskId, SId<User> userId));
```

Comment:
Are we okay with raising domain events not from aggregate roots?

Reply:
Yes, at least that is what I did for the last couple of years ^^'